File: TreeBreadthFirstSearch/Java/EX4_FindLevelAverages.java
package TreeBreadthFirstSearch;

import java.util.*;
//Given a binary tree, populate an array to represent the averages of all of its levels.
public class EX4_FindLevelAverages {
    /*Algorithm
    1. add root node to queue
    2. while queue is not empty
    3. for size of que
      3a. poll next item in quen
      3b. add val to sum
      3c. add left and right to queue
     4. add sum to results
     5. return results when queue is empty
     */

    public static List<Double> findLevelAverages(TreeNode root){
        List<Double> results = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        double sum;
        int size;
        while(!queue.isEmpty()){
          sum = 0;
          size = queue.size();
            for (int i = 0; i < size ; i++) {
                TreeNode curr = queue.poll();
                sum += curr.val;
                if (curr.right != null){
                    queue.offer(curr.right);
                }
                if (curr.left != null){
                    queue.offer(curr.left);
                }
            }
            results.add(sum/size);
        }
        return results;
    }
    public static void main(String[] args) {
        TreeNode root = new TreeNode(12);
        root.left = new TreeNode(7);
        root.right = new TreeNode(1);
        root.left.left = new TreeNode(9);
        root.left.right = new TreeNode(2);
        root.right.left = new TreeNode(10);
        root.right.right = new TreeNode(5);
        List<Double> result = findLevelAverages(root);
        System.out.print("Level averages are: " + result);
    }
}

File: TreeBreadthFirstSearch/Python/EX4_FindLevelAverages.py
# //Given a binary tree, populate an array to represent the averages of all of its levels.
from collections import deque


class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None


# Algorithm
#     1. add root node to queue
#     2. while queue is not empty
#     3. for size of que
#       3a. poll next item in queue
#       3b. add val to sum
#       3c. add left and right to queue
#      4. add sum to results
#      5. return results when queue is empty

def find_level_averages(root):
    results = []
    queue = deque()
    queue.append(root)
    while queue:
        total = 0
        size = len(queue)
        for i in range(size):
            curr = queue.popleft()
            total += curr.val
            if curr.left:
                queue.append(curr.left)
            if curr.right:
                queue.append(curr.right)
        results.append(sum)
    return results


def main():
    root = TreeNode(12)
    root.left = TreeNode(7)
    root.right = TreeNode(1)
    root.left.left = TreeNode(9)
    root.left.right = TreeNode(2)
    root.right.left = TreeNode(10)
    root.right.right = TreeNode(5)
    print("Level averages are: " + str(find_level_averages(root)))


main()
